digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/NodesAndJunctionsService.scala

```diff
@@ -496,90 +497,86 @@ class NodesAndJunctionsService(roadwayDAO: RoadwayDAO, roadwayPointDAO: RoadwayP
       2.2)  and at the beginning/end of each road part, ely borders, or when road type changes
       2.3)  on each junction with a road number (except number over 70 000)
      */
-  def handleNodePointTemplates(roadwayChanges: List[ProjectRoadwayChange], projectLinks: Seq[ProjectLink], mappedRoadwayNumbers: Seq[ProjectRoadLinkChange], username: String = "-"): Unit = {
+  def handleNodePoints(roadwayChanges: List[ProjectRoadwayChange], projectLinks: Seq[ProjectLink], mappedRoadwayNumbers: Seq[ProjectRoadLinkChange], username: String = "-"): Unit = {
+    @tailrec
+    def continuousNodeSections(seq: Seq[ProjectLink], roadTypesSection: Seq[Seq[ProjectLink]] = Seq.empty[Seq[ProjectLink]]): (Seq[ProjectLink], Seq[Seq[ProjectLink]]) = {
+      if (seq.isEmpty) {
+        (Seq(), roadTypesSection)
+      } else {
+        val roadType = seq.headOption.map(_.roadType.value).getOrElse(0)
+        val continuousProjectLinks = seq.takeWhile(pl => pl.roadType.value == roadType)
+        continuousNodeSections(seq.drop(continuousProjectLinks.size), roadTypesSection :+ continuousProjectLinks)
+      }
+    }
+
+    def createNodePointIfNeeded(projectLink: ProjectLink, addrM: Long, pos: BeforeAfter, reversed: Boolean, currentNodePoint: Option[NodePoint]): Unit = {
+      val roadwayPointId = roadwayPointDAO.fetch(projectLink.roadwayNumber, addrM).map(_.id)
+        .getOrElse(roadwayPointDAO.create(projectLink.roadwayNumber, addrM, username))
+
+      currentNodePoint match {
+        case Some(np) =>
+          val nodePointId = if (np.nodePointType == CalculatedNodePoint) Some(NewIdValue) else if (reversed) Some(np.id) else None
+          nodePointId.map { id =>
+            if (id == np.id) {
+              logger.info(s"Updated node point $id, due to reverse operation")
+              nodePointDAO.update(Seq(np.copy(beforeAfter = BeforeAfter.switch(np.beforeAfter))))
+            } else {
+              logger.info(s"Expired calculated node point ${np.id}")
```

Comment:
I do not understand why CalculatedNodePoint is copied as RoadNodePoint. I think this should not be done.  Definition of CalculatedNodePoint can be found from VIITE-1862

Reply:
That was the point of this fix. If you check `VIITE-2398 Jiira task` you'll notice that yes, there was a Calculated Node Point that should be modified to be Road Node Point.

Reply:
Calculated nodepoint is calculated as an average from junction m-values. I think Road node point is calculated with different algoritm. Because of that I think Road node point should be created as if there was no calculated node point. Now there is double logic for creating calculated node point. Original implementation takes care of calculated nodepoint both in project and in node tool. This new implementation does it only for project.